src/main/java/org/tikv/common/importer/ImporterClient.java

```diff
@@ -0,0 +1,191 @@
+      ArrayList<ImportSstpb.Pair> pairs = new ArrayList<>(maxKVBatchSize);
+      for (int i = 0; i < maxKVBatchSize; i++) {
+        if (iterator.hasNext()) {
+          Pair<ByteString, ByteString> pair = iterator.next();
+          pairs.add(ImportSstpb.Pair.newBuilder().setKey(pair.first).setValue(pair.second).build());
+          totalBytes += (pair.first.size() + pair.second.size());
+        }
+        if (totalBytes > maxKVBatchBytes) {
+          break;
+        }
+      }
+      if (!openStream) {
+        init();
+        startRawWrite();
+        rawWriteMeta();
+        openStream = true;
+      }
+      rawWriteBatch(pairs);
+    }
+
+    if (openStream) {
+      finishRawWrite();
+      ingest();
+    }
+  }
+
+  private void init() {
+    long regionId = region.getId();
+    Metapb.RegionEpoch regionEpoch = region.getRegionEpoch();
+    ImportSstpb.Range range =
+        ImportSstpb.Range.newBuilder()
+            .setStart(minKey.toByteString())
+            .setEnd(maxKey.toByteString())
+            .build();
+
+    sstMeta =
+        ImportSstpb.SSTMeta.newBuilder()
+            .setUuid(uuid)
+            .setRegionId(regionId)
+            .setRegionEpoch(regionEpoch)
+            .setRange(range)
+            .build();
+
+    clientList = new ArrayList<>();
+    for (Metapb.Peer peer : region.getPeersList()) {
+      long storeId = peer.getStoreId();
+      TiStore store = tiSession.getRegionManager().getStoreById(storeId);
+      ImporterStoreClient importerStoreClient =
+          tiSession.getImporterRegionStoreClientBuilder().build(store);
+      clientList.add(importerStoreClient);
+
+      if (region.getLeader().getStoreId() == storeId) {
+        clientLeader = importerStoreClient;
+      }
+    }
+  }
+
+  private void startRawWrite() {
+    for (ImporterStoreClient client : clientList) {
+      client.startRawWrite();
+    }
+  }
+
+  private void rawWriteMeta() {
+    ImportSstpb.RawWriteRequest request =
+        ImportSstpb.RawWriteRequest.newBuilder().setMeta(sstMeta).build();
+    for (ImporterStoreClient client : clientList) {
+      client.rawWriteBatch(request);
+    }
+  }
+
+  private void rawWriteBatch(List<ImportSstpb.Pair> pairs) {
+    ImportSstpb.RawWriteBatch batch;
+
+    if (ttl == null || ttl <= 0) {
+      batch = ImportSstpb.RawWriteBatch.newBuilder().addAllPairs(pairs).build();
+    } else {
+      batch = ImportSstpb.RawWriteBatch.newBuilder().addAllPairs(pairs).setTtl(ttl).build();
+    }
+
+    ImportSstpb.RawWriteRequest request =
+        ImportSstpb.RawWriteRequest.newBuilder().setBatch(batch).build();
+    for (ImporterStoreClient client : clientList) {
+      client.rawWriteBatch(request);
+    }
+  }
+
+  private void finishRawWrite() {
+    for (ImporterStoreClient client : clientList) {
+      client.finishRawWrite();
+    }
+  }
+
+  private void ingest() throws GrpcException {
+    int returnNumber = 0;
+    while (returnNumber < clientList.size()) {
```

Comment:
Will this dead loop if a client never succeed?

Reply:
then will go to `client.hasRawWriteResponseError()` and `throw new GrpcException`